File: analysis/import/SCMLogParser/src/test/java/de/maibornwolff/codecharta/model/input/VersionControlledFileTest.java
package de.maibornwolff.codecharta.model.input;

import de.maibornwolff.codecharta.model.input.metrics.CommitMetric;
import de.maibornwolff.codecharta.model.input.metrics.MetricsFactory;
import de.maibornwolff.codecharta.model.input.metrics.ModificationMetric;
import org.junit.Test;

import java.time.LocalDateTime;
import java.util.Arrays;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.*;

public class VersionControlledFileTest {

    private final MetricsFactory metricsFactory = mock(MetricsFactory.class);

    @Test
    public void versionControlledFileHoldsInitallyOnlyTheFilename() {
        // given
        String filename = "filename";

        // when
        VersionControlledFile versionControlledFile = new VersionControlledFile(filename, metricsFactory);

        // then
        assertThat(versionControlledFile.getFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.getAuthors()).isEmpty();
    }


    @Test
    public void ignoresCommitsForDifferentFiles() {
        // given
        ModificationMetric modificationMetric = mock(ModificationMetric.class);
        CommitMetric commitMetric = mock(CommitMetric.class);

        VersionControlledFile versionControlledFile = new VersionControlledFile(
                "filename",
                Arrays.asList(modificationMetric),
                Arrays.asList(commitMetric)
        );

        // when
        Modification modification = new Modification("anotherFilename");
        Commit commit = createCommit("An Author", modification);
        versionControlledFile.registerCommit(commit);

        // then
        assertThat(versionControlledFile.getAuthors()).isEmpty();

        verify(modificationMetric, times(0)).registerModification(any());

        verify(commitMetric, times(0)).registerCommit(any());
    }

    @Test
    public void canRegisterACommit() {
        // given
        ModificationMetric modificationMetric = mock(ModificationMetric.class);
        CommitMetric commitMetric = mock(CommitMetric.class);

        String filename = "filename";
        String author = "An Author";
        VersionControlledFile versionControlledFile = new VersionControlledFile(
                filename,
                Arrays.asList(modificationMetric),
                Arrays.asList(commitMetric)
        );

        // when
        Modification modification = new Modification(filename);
        Commit commit = createCommit(author, modification);
        versionControlledFile.registerCommit(commit);

        // then
        assertThat(versionControlledFile.getFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.getAuthors()).containsExactly(author);

        verify(modificationMetric, times(1)).registerModification(any());
        verify(modificationMetric, times(1)).registerModification(eq(modification));

        verify(commitMetric, times(1)).registerCommit(any());
        verify(commitMetric, times(1)).registerCommit(eq(commit));
    }

    @Test
    public void canRegisterMultipleCommitsWithSimpleStatistics() {
        // given
        ModificationMetric modificationMetric = mock(ModificationMetric.class);
        CommitMetric commitMetric = mock(CommitMetric.class);

        String filename = "filename";
        String author1 = "An Author";
        String author2 = "2nd Author";
        VersionControlledFile versionControlledFile = new VersionControlledFile(
                filename,
                Arrays.asList(modificationMetric),
                Arrays.asList(commitMetric)
        );

        // when
        Modification modification1 = new Modification(filename);
        Commit commit1 = createCommit(author1, modification1);
        versionControlledFile.registerCommit(commit1);
        Modification modification2 = new Modification(filename);
        Commit commit2 = createCommit(author2, modification2);
        versionControlledFile.registerCommit(commit2);
        Modification modification3 = new Modification(filename);
        Commit commit3 = createCommit(author1, modification3);
        versionControlledFile.registerCommit(commit3);

        // then
        assertThat(versionControlledFile.getFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.getAuthors()).containsExactlyInAnyOrder(author1, author2);

        verify(commitMetric, times(3)).registerCommit(any());
        verify(commitMetric, times(1)).registerCommit(eq(commit1));
        verify(commitMetric, times(1)).registerCommit(eq(commit2));
        verify(commitMetric, times(1)).registerCommit(eq(commit3));

        verify(modificationMetric, times(3)).registerModification(any());
        verify(modificationMetric, times(1)).registerModification(eq(modification1));
        verify(modificationMetric, times(1)).registerModification(eq(modification2));
        verify(modificationMetric, times(1)).registerModification(eq(modification3));
    }

    private Commit createCommit(String author, Modification modification) {
        return new Commit(author, Arrays.asList(modification), LocalDateTime.now());
    }

}

File: analysis/import/SCMLogParserV2/src/test/kotlin/de/maibornwolff/codecharta/importer/scmlogparserv2/input/VersionControlledFileTest.kt
package de.maibornwolff.codecharta.importer.scmlogparserv2.input

import de.maibornwolff.codecharta.importer.scmlogparserv2.input.metrics.Metric
import de.maibornwolff.codecharta.importer.scmlogparserv2.input.metrics.MetricsFactory
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.time.OffsetDateTime
import java.util.Arrays

class VersionControlledFileTest {

    @Test
    fun versionControlledFileHoldsInitallyOnlyTheFilename() {
        // given
        val metricsFactory = mockk<MetricsFactory>(relaxed = true)
        val filename = "filename"

        // when
        val versionControlledFile = VersionControlledFile(filename, metricsFactory)

        // then
        assertThat(versionControlledFile.filename).isEqualTo(filename)
        assertThat(versionControlledFile.authors).isEmpty()
    }

    @Test
    fun canRegisterMetricsByMetricsFactory() {
        // given
        val metricName = "metric"
        val metric = mockk<Metric>()
        every { metric.metricName() } returns metricName
        every { metric.value() } returns 1

        val metricsFactory = mockk<MetricsFactory>()
        every { metricsFactory.createMetrics() } returns Arrays.asList(metric)

        val versionControlledFile = VersionControlledFile(
            "filename",
            metricsFactory
        )

        // when
        val metricsMap = versionControlledFile.metricsMap

        // then
        assertThat(metricsMap).hasSize(1)
        assertThat(versionControlledFile.getMetricValue(metricName))
            .isEqualTo(1)
    }

    @Test
    fun canRegisterASimpleCommit() {
        // given
        val modificationMetric = mockk<Metric>(relaxed = true)

        val filename = "filename"
        val author = "An Author"
        val versionControlledFile = VersionControlledFile(
            filename,
            Arrays.asList(modificationMetric)
        )

        // when
        val modification = Modification(filename)
        val commit = createCommit(author, modification)
        versionControlledFile.registerCommit(commit, modification)

        // then
        assertThat(versionControlledFile.filename).isEqualTo(filename)
        assertThat(versionControlledFile.authors).containsExactly(author)

        verify(exactly = 1) { modificationMetric.registerModification(any()) }
        verify(exactly = 1) { modificationMetric.registerModification(modification) }
    }

    private fun createCommit(author: String, modification: Modification): Commit {
        return Commit(author, Arrays.asList(modification), OffsetDateTime.now())
    }
}
